File: queue_linkedlist.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.util.LinkedList;
import java.util.Queue;
import com.sun.speech.freetts.*;

public class queue_linkedlist {

    static class Antrian {
        int nomor;
        String nama;

        Antrian(int nomor, String nama) {
            this.nomor = nomor;
            this.nama = nama;
        }

        @Override
        public String toString() {
            return "No: " + nomor + " | Nama: " + nama;
        }

    }

    private Queue<Antrian> queue = new LinkedList<>();
    private int nomorCounter = 1;

    private JTextArea textArea;
    private JLabel labelPanggilan;

    public queue_linkedlist() {
        JFrame frame = new JFrame("Simulasi Antrian Bank");
        frame.setSize(500, 400);
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setLayout(new BorderLayout());

        // Text Area
        textArea = new JTextArea();
        textArea.setEditable(false);
        frame.add(new JScrollPane(textArea), BorderLayout.CENTER);

        // Panel Button
        JPanel panel = new JPanel();

        JButton btnAmbil = new JButton("Ambil Antrian");
        JButton btnTampil = new JButton("Tampilkan Antrian");
        JButton btnPanggil = new JButton("Panggil Antrian");

        panel.add(btnAmbil);
        panel.add(btnTampil);
        panel.add(btnPanggil);

        frame.add(panel, BorderLayout.SOUTH);

        // Label atas
        labelPanggilan = new JLabel("Belum ada panggilan", JLabel.CENTER);
        labelPanggilan.setFont(new Font("Arial", Font.BOLD, 16));
        frame.add(labelPanggilan, BorderLayout.NORTH);

        // Event Ambil Antrian
        btnAmbil.addActionListener(e -> {
            String nama = JOptionPane.showInputDialog(frame, "Masukkan Nama:");
            if (nama != null && !nama.trim().isEmpty()) {
                Antrian a = new Antrian(nomorCounter++, nama);
                queue.add(a);
                JOptionPane.showMessageDialog(frame, "Nomor Antrian Anda: " + a.nomor);
            }
        });

        // Event Tampilkan
        btnTampil.addActionListener(e -> {
            textArea.setText("");
            if (queue.isEmpty()) {
                textArea.setText("Antrian kosong");
            } else {
                for (Antrian a : queue) {
                    textArea.append(a.toString() + "\n");
                }
            }
        });

        // Event Panggil
        btnPanggil.addActionListener(e -> {
            if (queue.isEmpty()) {
                labelPanggilan.setText("Tidak ada antrian");
                return;
            }

            Antrian a = queue.poll();
            String panggilan = "Memanggil nomor " + a.nomor + ", atas nama " + a.nama;
            labelPanggilan.setText(panggilan);

            // Voice sederhana (pakai beep)
          speak("Nomor " + a.nomor + ", atas nama " + a.nama + ",  please go to counter");
        });

        frame.setVisible(true);
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(queue_linkedlist::new);
    }

    public void speak(String text) {
    System.setProperty("freetts.voices",
        "com.sun.speech.freetts.en.us.cmu_us_kal.KevinVoiceDirectory");

    VoiceManager vm = VoiceManager.getInstance();
    Voice voice = vm.getVoice("kevin16");

    if (voice == null) {
        System.out.println("Voice tidak ditemukan!");
        return;
    }

    voice.allocate();
    voice.speak(text);
    voice.deallocate();
}
}

File: queue_linkedlist.py
import tkinter as tk
from tkinter import simpledialog, messagebox
from collections import deque
from gtts import gTTS
from playsound import playsound
import os




def speak(text):
    tts = gTTS(text=text, lang='id')  # bahasa Indonesia
    tts.save("suara.mp3")
    playsound("suara.mp3")
    os.remove("suara.mp3")  # hapus biar gak numpuk

class Antrian:
    def __init__(self, nomor, nama):
        self.nomor = nomor
        self.nama = nama

    def __str__(self):
        return f"No: {self.nomor} | Nama: {self.nama}"

class QueueApp:
    def __init__(self, root):
        self.root = root
        self.root.title("Simulasi Antrian Bank")
        self.root.geometry("500x400")

        self.queue = deque()
        self.nomor_counter = 1

        # Label atas
        self.label = tk.Label(root, text="Belum ada panggilan", font=("Arial", 14))
        self.label.pack(pady=10)

        # Text area
        self.text_area = tk.Text(root, height=10)
        self.text_area.pack(padx=10, pady=10)

        # Frame tombol
        frame = tk.Frame(root)
        frame.pack()

        btn_ambil = tk.Button(frame, text="Ambil Antrian", command=self.ambil_antrian)
        btn_tampil = tk.Button(frame, text="Tampilkan Antrian", command=self.tampilkan_antrian)
        btn_panggil = tk.Button(frame, text="Panggil Antrian", command=self.panggil_antrian)

        btn_ambil.grid(row=0, column=0, padx=5)
        btn_tampil.grid(row=0, column=1, padx=5)
        btn_panggil.grid(row=0, column=2, padx=5)

    def ambil_antrian(self):
        nama = simpledialog.askstring("Input", "Masukkan Nama:")
        if nama:
            a = Antrian(self.nomor_counter, nama)
            self.queue.append(a)
            messagebox.showinfo("Info", f"Nomor Antrian Anda: {a.nomor}")
            self.nomor_counter += 1

    def tampilkan_antrian(self):
        self.text_area.delete("1.0", tk.END)
        if not self.queue:
            self.text_area.insert(tk.END, "Antrian kosong")
        else:
            for a in self.queue:
                self.text_area.insert(tk.END, str(a) + "\n")

    def panggil_antrian(self):
        if not self.queue:
            self.label.config(text="Tidak ada antrian")
            return

        a = self.queue.popleft()
        teks = f"Memanggil nomor {a.nomor}, atas nama {a.nama}"
        self.label.config(text=teks)

        # 🔊 Voice
        speak(f"Nomor {a.nomor}, {a.nama}, silahkan menuju loket")

# Run aplikasi
root = tk.Tk()
app = QueueApp(root)
root.mainloop()